### PR TITLE
Use Golang 1.18 for Gitea 1.17 release

### LINKED ISSUE #19917: [Proposal] Use Golang 1.18 for Gitea 1.17 release
Use Golang 1.18 (as minimal requirement) for Gitea 1.17 release, make sure the Golang version is still actively supported during Gitea 1.17 lifecycle.

And Gitea can start using Golang generics from then on.

### PR DESCRIPTION
**Pull request by wxiaoguang**
Use Golang 1.18 (as minimal requirement) for Gitea 1.17 release, make sure the Golang version is still actively supported during Gitea 1.17 lifecycle.

~~And Gitea can start using Golang generics from then on.~~
update: As discussed in https://github.com/go-gitea/gitea/issues/19917#issuecomment-1149636662 , only use it when it's carefully designed.

----

(sorry for that I opened a similar issue and opened this PR, I was going to ask to delete the issue, but soon it contains more important information 😂)

### CODE REVIEW

**Review by zeripath — APPROVED** [review]
Agreed with using 1.18 - Disagree with adding generics directly to gitea1.17. 

We need to get 1.17 ready for release - adding generics refactors is going to delay that. 

Get RC1 out then we can think about adding in Generics. (But please be aware of the caveats with their current implementation in the golang runtime - They are not free and may come with serious costs where interface{} is used.)

**Review by 42wim — APPROVED** [review]

**Review by delvh — APPROVED** [review]

**Comment by delvh** [comment]
@zeripath What about releasing 1.17 as soon as possible and then backporting refactoring PRs once they appear?

**Comment by zeripath** [comment]
> @zeripath What about releasing 1.17 as soon as possible and then backporting refactoring PRs once they appear?

Generally we should avoid backporting refactoring PRs as they often come with bugs and or change of behaviour. However, I think if there were something that would need backporting that included generics then having 1.17 run go1.18 would allow us to consider backporting those generics to avoid having to completely restructure that backport if necessary. I expect most of the time though, that making a non-generic backport would be easy so we shouldn't need to backport generics.

If it were getting too difficult to backport things we should just consider releasing 1.18 quickly as really our releases are too long and too big and we're making our lives difficult because of this.

**Comment by delvh** [comment]
> If it were getting too difficult to backport things we should just consider releasing 1.18 quickly as really our releases are too long and too big and we're making our lives difficult because of this.

Strongly agree, but that's a completely different topic that we should tackle soon.

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 1]
```diff
@@ -315,8 +315,8 @@ The following configuration set `Content-Type: application/vnd.android.package-a
 - `SSL_MAX_VERSION`: **\<empty\>**: Set the maximum version of ssl support.
 - `SSL_CURVE_PREFERENCES`: **X25519,P256**: Set the preferred curves,
 - `SSL_CIPHER_SUITES`: **ecdhe_ecdsa_with_aes_256_gcm_sha384,ecdhe_rsa_with_aes_256_gcm_sha384,ecdhe_ecdsa_with_aes_128_gcm_sha256,ecdhe_rsa_with_aes_128_gcm_sha256,ecdhe_ecdsa_with_chacha20_poly1305,ecdhe_rsa_with_chacha20_poly1305**: Set the preferred cipher suites.
-  - If there is not hardware support for AES suites by default the cha cha suites will be preferred over the AES suites
-  - supported suites as of go 1.17 are:
+  - If there is no hardware support for AES suites, by default the ChaCha suites will be preferred over the AES suites.
+  - supported suites as of Go >= 1.17 are:
```
Comment (lunny) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]: 1.18？

Comment (wxiaoguang) [reply docs/content/doc/advanced/config-cheat-sheet.en-us.md]:
```suggestion
  - supported suites as of Go 1.18 are:
```

**Comment by lafriks** [comment]
shouldn't go mod tidy also changed without compat 1.17 flag than?

**Comment by wxiaoguang** [comment]
> shouldn't go mod tidy also changed without compat 1.17 flag than?

Do you mean these code?

```
.PHONY: tidy
tidy:
	$(eval MIN_GO_VERSION := $(shell grep -Eo '^go\s+[0-9]+\.[0-9.]+' go.mod | cut -d' ' -f2))
	$(GO) mod tidy -compat=$(MIN_GO_VERSION)
```

It's already automated to parse `go.mod`.

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
I hope all package maintainer do have enough time to prepare for that ...
... but it's a good move to use new libs who do relay on generics :)

**Comment by wxiaoguang** [comment]
> I hope all package maintainer do have enough time to prepare for that ... ... but it's a good move to use new libs who do relay on generics :)

Recently I wrote another `CopyMap` in `#17624`, then I remembered the generic again 😂

**Pull request merged**